### DISCUSSION TITLE
Cardinality analysis per tenant

### DISCUSSION THREAD

**Discussion by kazeborja** [grafana/mimir#4590] [Help and support] [answered]
Hi all,

Reading about cardinality analysis for Mimir it always points to these set of [amazing dashboards](https://grafana.com/docs/grafana-cloud/billing-and-usage/control-prometheus-metrics-usage/cardinality-management/) that are meant for GEM.

I've been trying to recreate them using direct Prometheus queries, but as stated everywhere getting things like the number of metrics/series/labels or label_values is quite heavy and something I'd love not to have in a dashboard directly... Which I guess explains why for the GEM dashboards this points to another database.

From a lot of reading I have ended up enabling the "cardinality_analysis_enabled" and got access for the endpoints which are fine for investigations of the cardinality, but not so nice to show some dashboards with the usage, so I was wondering if there's any recommended way to get these kind of metrics directly from the "tsdb" (similar to Prometheus) so I can store and visualise them.

The closest thing I was able to find is the "cortex_ingester_active_series" metric that indeed sorts out the counting for the series :)

Thanks a lot in advance for the help

**Comment by wilfriedroset** [comment, marked as answer]
Hello, you are correct. you need to enable cardinality analysis as a first step to recreate the dashboard.
Second you need to have a way to query the cardinality API from grafana. This can be solved by [json api datasource](https://grafana.github.io/grafana-json-datasource/). With both the datasource and the API you can recreate the dashboard.
All three dashboards relies on `/prometheus/api/v1/cardinality/label_names` and `/prometheus/api/v1/cardinality/label_values`

**Reply by kazeborja** [reply]
This indeed put me in the right direction, thanks a lot for the hint, to learn crazy JSONata we go :)

**Reply by jwdalton-f** [reply]
@kazeborja Were you able to make any progress in creating these? I was also thinking about doing this, and would greatly appreciate if you have any work on this that you can share!

**Reply by kazeborja** [reply]
@jwdalton-f yup, I made them work in a pretty similar way as the ones mentioned in the documentation, will be happy to share them with some instructions (please note currently I'm running them on top of Grafana8.5 and specifically the JSON api datasource has undergo some changes, so it might be slightly different if you use a newer version of Grafana).

Let me know how can I share the dashboards with you :)

**Reply by jwdalton-f** [reply]
@kazeborja That would be amazing! Can you export the dashboards as json and send them to joseph.dalton@flexe.com?

**Reply by akselleirv** [reply]
@kazeborja would you mind sharing the dashboard here? It would be greatly appreciated!

**Reply by kazeborja** [reply]
No problem, let me prepare some repo with the JSONs inside and some instructions, will be back

**Reply by akselleirv** [reply]
That would be great, thanks!

**Reply by akselleirv** [reply]
@kazeborja just a friendly reminder in case you have forgotten it. Would also be very helpful if you could just send the dashboard, and then I can try to figure out the rest myself.

**Comment by kazeborja** [comment]
Sorry for the late reply, but bureaucracy :), please find the dashboards and some simple usage instructions here https://github.com/cerndb/grafana-mimir-cardinality-dashboards/tree/main

**Reply by clouedoc** [reply]
Thanks a lot 💯

**Reply by akselleirv** [reply]
Thank you!!

**Comment by michaelpeterswa** [comment]
I'm very glad there was some exploration here, we just started encountering cardinality issues in our deployment and we looked for a source of the dashboard + config info only to find ~~it's locked behind Grafana Cloud/GEM...~~ (**_Edit: looks like it's publicly available, on play.grafana.org_**)  I don't think it's immediately clear in the discussion above, but we only needed one "JSON API Datasource" and the dashboards append the `/label_names` and `/label_values`. Given that, our final URL for the JSON Datasource ended up being `http://***.***.svc.cluster.local/prometheus/api/v1/cardinality`. Hopefully that helps the next person who looks. Thanks @kazeborja and @wilfriedroset. 😄

**Reply by michaelpeterswa** [reply]
In addition, if the JSON ever changes, a more updated version can be found at each of the dashboards listed here: https://play.grafana.org/dashboards?query=cardinality
